hudi-client/hudi-client-common/src/main/java/org/apache/hudi/keygen/ComplexAvroKeyGenerator.java

```diff
@@ -44,6 +48,9 @@ public ComplexAvroKeyGenerator(TypedProperties props) {
 
   @Override
   public String getRecordKey(GenericRecord record) {
+    if (autoGenerateRecordKeys()) {
+      return StringUtils.EMPTY_STRING;
+    }
```

Comment:
Is it safe to use the empty string, the payloads can be merged while going through the merging procedure.

Reply:
this is kind of unavoidable as of current structure. For eg, even to fetch partition path, our KeyGenerator interface, only exposes 
```
HoodieKey getKey(GenericRecord record)
```

So, to fetch partition path, we have to call getKey(genRec).getPartitionPath and hence I had to return empty string here. we don't want to add a new api to the interface just for this purpose. 

Incase of auto key gen flows, we generate the record keys explicitly (not via key gen class) and add it to HoodieKey that we materialize in memory for all records.

I can sync up w/ you f2f to clarify this. 
Ideally, we need to have 2 different interfaces. one to generate partition path and one to generate record key. and so some of these workarounds may not be required. but w/ current structure, we use a single key gen class to generate both record keys and partition paths as well.

Reply:
We already have `getRecordKey` and `getPartitionPath` as the public API, if you want to fix the `HoodieKey`, shouldn't the `HoodieKey getKey()` be fixed instead?

Reply:
I'm wondering whether we can add a new specific key generator for the keyless scenario, this new key generator is a wrapper for the existing key generator, and it masks out some invocations like:

1. `#getRecordKey`, throws exception directly
2. `#getKey`, returns a empty string for the HoodieKey record key

This could avoid to introduce potential bugs and is very straight-forward.

Reply:
synced up directly. we will take it up in a follow up patch where we will de-couple partition path generation and record key generation 
Refer to https://github.com/apache/hudi/pull/7699
[HUDI-5535](https://issues.apache.org/jira/browse/HUDI-5535).

Reply:
Agree, we can take a refactoring in another patch for separating record key and partition path generation. What I'm suggesting is orthogonal to the refactoring and it minimize the changes, can we do that first in this patch?

Reply:
Do we have alignment on the wrapping KeyGenerator? @nsivabalan what @danny0405  is saying is - have a `HoodieInternalKeyGenerator` class and do all these checks in that one file, while delegating to the inner key generator. Makes sense to me, will avoid touching all key generators. no?